File: app/src/main/java/com/example/rssreader/presentation/sourcelist/SourceAdapter.java
package com.example.rssreader.presentation.sourcelist;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.rssreader.R;
import com.example.rssreader.data.RssSource;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SourceAdapter extends RecyclerView.Adapter<SourceAdapter.ViewHolder> {
    public interface OnItemClickListener {
        void onClick(int position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.name)
        public TextView name;

        @BindView(R.id.url)
        public TextView url;

        public ViewHolder(@NonNull View item) {
            super(item);
            ButterKnife.bind(this, item);
            item.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            if (clickListener == null || getAdapterPosition() == RecyclerView.NO_POSITION) {
                return;
            }
            clickListener.onClick(getAdapterPosition());
        }
    }

    @NonNull
    private List<RssSource> data = new ArrayList<>();

    @Nullable
    private OnItemClickListener clickListener;

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.item_source, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        RssSource source = data.get(position);
        holder.name.setText(source.getName());
        holder.url.setText(source.getUrl());
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @NonNull
    public List<RssSource> getData() {
        return data;
    }

    public void setData(@NonNull List<RssSource> data) {
        this.data = data;
    }

    public void setOnItemClickListener(@Nullable OnItemClickListener listener) {
        clickListener = listener;
    }
}

File: app/src/main/java/com/example/rssreader/presentation/sourcelist/SourceAdapter.kt
package com.example.rssreader.presentation.sourcelist

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.rssreader.R
import com.example.rssreader.domain.entity.RssSource
import com.example.rssreader.presentation.BaseAdapter
import com.example.rssreader.presentation.BaseViewHolder
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_source.*
import kotlinx.android.synthetic.main.item_source.view.*

class SourceAdapter : BaseAdapter<SourceAdapter.ViewHolder>() {
    var onRemoveClickListener: (position: Int) -> Unit = {}

    inner class ViewHolder(
        override val containerView: View,
        itemClickListener: (Int) -> Unit,
        removeClickListener: (Int) -> Unit
    ) : BaseViewHolder(containerView, itemClickListener), LayoutContainer {

        init {
            containerView.remove.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    removeClickListener(adapterPosition)
                }
            }
        }
    }

    var data: List<RssSource> = emptyList()
        set(value) {
            val result = DiffUtil.calculateDiff(DiffCallback(field, value))
            field = value
            result.dispatchUpdatesTo(this)
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflate(R.layout.item_source, parent)
        return ViewHolder(view, onItemClickListener, onRemoveClickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val source = data[position]
        holder.name.text = source.name
        holder.url.text = source.url
    }

    override fun getItemCount(): Int = data.size
}

class DiffCallback(
    private val old: List<RssSource>,
    private val new: List<RssSource>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = old.size

    override fun getNewListSize(): Int = new.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        old[oldItemPosition].id == new[newItemPosition].id

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        old[oldItemPosition] == new[newItemPosition]
}